refactor(dev-infra): remove usages and dependency on shelljs

josephperrott opened the pull request:
Remove usages of shelljs and instead use spawn/spawnSync.

josephperrott committed:
refactor(dev-infra): remove usages and dependency on shelljs

Remove usages of shelljs and instead use spawn/spawnSync.

gkalpak commented on dev-infra/format/formatters/prettier.ts:
```diff
@@ -27,8 +27,11 @@ export class Prettier extends Formatter {
-      this.config['prettier'] ? exec(`${this.binaryFilePath} --find-config-path .`).trim() : '';
+  private configPath = this.config['prettier'] ?
+      spawnSync(this.binaryFilePath, ['--find-config-path', '.'], {
+        encoding: 'utf8'
```
You probably want to use `shell: true`, otherwise this will not work on Windows unless `binaryFilePath` contains the `.cmd` extension:

```suggestion
        encoding: 'utf8',
        shell: true,
```

gkalpak reviewed (commented):
Just a note of caution: Making `spawn()`/`exec()` work correctly across platforms can be quite tricky (as I've found out the hard way :grin:). ShellJS can help avoid headaches.

But since our use of `spawn()` is limited, we could probably get away (and just fix things as we find out they're broken) :grin:

It might still be worth exposing an `exec()` utility (to take care of `encoding: 'utf8'` and `shell: true` in one place).

devversion reviewed (approved):
LGTM in general. I agree with @gkalpak that a common utility at least makes sense (for `spawn`/`exec`). We might be able to always use the existing `spawnWithDebugOutput` utility function here?

josephperrott commented:
@devversion @gkalpak PTAL, I added a common utility to do synchronous spawn calls.

gkalpak commented on dev-infra/format/formatters/prettier.ts:
```diff
@@ -7,7 +7,7 @@
 
 import {join} from 'path';
-import {exec} from 'shelljs';
+import {spawnSyncWithDebugOutput} from '../../utils/child-process';
 
```
```suggestion

import {spawnSyncWithDebugOutput} from '../../utils/child-process';
```

devversion commented on dev-infra/utils/child-process.ts:
```diff
@@ -93,10 +101,27 @@ export function spawnWithDebugOutput(
+    command: string, args: string[],
+    options: SpawnSyncWithDebugOutputOptions = {}): SpawnProcessResult {
+  const childProcess =
+      spawnSync(command, args, {...options, encoding: 'utf8', shell: true, stdio: 'pipe'});
```
I think one significant difference with the async variant above (which is why we created it), is that stdout is printed to stderr. The reason for this is that we want to have the ability to keep stdout clean for tools that return e.g.  a JSON to stdout (e.g. when `yarn ng-dev release build --json`) runs.

I'd propose we do the same here too and support the `mode` property as above? WDYT?

gkalpak replied:
+1

I also think it might make sense to preserve other semantics of the async method, such as throwing an error (which is the equivalent of rejecting a promise) when `status !== 0`.

josephperrott replied:
We can, however there is not a great way of merging `stdout` and `stderr` together.  With async, the events are sent in order so we coalesce everything together in order.  With `spawnSync`, we only have the resulting output in `stdout` and `stderr`, without a knowledge of the order of each line output.

devversion replied:
That's fair. I guess you could create a stream and pass it as `stdio` (this is like the default pipe, but combined). Not sure if that is worth the effort though.

josephperrott replied:
Its possible, but you can't just create a generic stream. 

From [options.stdio](https://nodejs.org/api/child_process.html#child_process_options_stdio)
>
```
<Stream> object: Share a readable or writable stream that refers to a tty, file, socket, or a 
pipe with the child process. The stream's underlying file descriptor is duplicated in the child
process to the fd that corresponds to the index in the stdio array. The stream must have an
underlying descriptor (file streams do not until the 'open' event has occurred).
```


So the stream has to have references to a file in our case, and then we have to have the open method occur on the file descriptor, which is an async process.  So we can't ensure synchronous calls anymore.

devversion replied:
I see. good point. We should be okay keeping as is; but should rename the function to not suggest it's similar to the async `withDebugOutput` helper. We don't deal with any debug output here; rather it's a simple `spawnSync` wrapper with our Windows compatibility settings (e.g. `shell`).

devversion commented on dev-infra/utils/child-process.ts:
```diff
@@ -93,10 +101,27 @@ export function spawnWithDebugOutput(
 }
+
+/**
+ * Spawns a given command with the specified arguments inside a shell syncronously. All process
```
The comment here seems cut off.

gkalpak commented on dev-infra/format/run-commands-parallel.ts:
```diff
@@ -87,34 +87,33 @@ export function runFormatterInParallel(allFiles: string[], action: FormatterActi
-          resolve(failures);
-        }
-      });
+      spawnWithDebugOutput(spawnCmd, spawnArgs)
```
In contrast to what was happening before, the stdout and stderr output will be printed out (in addition to being returned). Is that desirable?

gkalpak commented on dev-infra/format/run-commands-parallel.ts:
```diff
@@ -87,34 +87,33 @@ export function runFormatterInParallel(allFiles: string[], action: FormatterActi
+          .then(
+              ({stdout, stderr, status: code}: SpawnProcessResult) => {
+                // Run the provided callback function.
+                const failed = formatter.callbackFor(action)(file, code, stdout, stderr);
```
Previously, this would be called on failure as well. Now it will only be called if `code === 0` (since otherwise the promise would be rejected). Is that intentional?

gkalpak commented on dev-infra/utils/child-process.ts:
```diff
@@ -6,21 +6,29 @@
 
 /** Interface describing the options for spawning a process. */
-export interface SpawnedProcessOptions extends Omit<SpawnOptions, 'stdio'> {
+export interface SpawnedProcessOptions extends Omit<SpawnOptions, 'stdio'|'shell'> {
```
Nit: 😇 
```suggestion
export interface SpawnedProcessOptions extends Omit<SpawnOptions, 'shell'|'stdio'> {
```

gkalpak commented on dev-infra/utils/child-process.ts:
```diff
@@ -6,21 +6,29 @@
+export type SpawnInteractiveCommandOptions = Omit<SpawnOptions, 'shell'|'stdio'>;
+
+/** Interface describing the options for spawning a syncronous process. */
+export type SpawnSyncWithDebugOutputOptions = Omit<SpawnSyncOptions, 'stdio'|'encoding'|'shell'>;
```
Nit: 😇 
```suggestion
export type SpawnSyncWithDebugOutputOptions = Omit<SpawnSyncOptions, 'encoding'|'shell'|'stdio'>;
```

gkalpak commented on dev-infra/utils/child-process.ts:
```diff
@@ -93,10 +101,27 @@ export function spawnWithDebugOutput(
       // and stdout log output if the output mode was set to `silent`.
       if (status === 0) {
-        resolve({stdout, stderr});
+        resolve({stdout, stderr, status});
```
So, if status is always going to be `0`, what is the benefit of including it in the return value? 🤔

josephperrott committed:
refactor(dev-infra): add spawnSync to child process utils, normalize naming of child-process utils

Create a `spawnSync` command for common usage, additionally update naming to use `spawn` instead of
`spawnWithDebugOutput`

gkalpak commented on dev-infra/release/stamping/env-stamp.ts:
```diff
@@ -66,18 +62,20 @@ function getSCMVersion(mode: EnvStampMode) {
 function getCurrentSha() {
-  return exec(`git rev-parse HEAD`);
+  const git = GitClient.get();
+  return git.getCurrentBranchOrRevision();
```
AFAICT, this does not return the SHA (but could return a branch name). If that is OK for our usecase here, we should update the function name and description to reflect what it actually does.

gkalpak commented on dev-infra/utils/child-process.ts:
```diff
@@ -45,18 +58,17 @@ export function spawnInteractiveCommand(
  *
  * @returns a Promise resolving with captured stdout and stderr on success. The promise
- *   rejects on command failure.
+ *   rejects on command failure
```
🤔

gkalpak commented on dev-infra/utils/child-process.ts:
```diff
@@ -83,20 +95,51 @@ export function spawnWithDebugOutput(
 }
+
+/**
+ * Spawns a given command with the specified arguments inside a shell syncronously.
```
```suggestion
 * Spawns a given command with the specified arguments inside a shell synchronously.
```

gkalpak commented on dev-infra/utils/child-process.ts:
```diff
@@ -83,20 +95,51 @@ export function spawnWithDebugOutput(
+ * @returns The command's stdout and stderr.
+ */
+export function spawnSync(
+    command: string, args: string[], options: SpawnOptions = {}): SpawnResult {
```
Shouldn't this be:

```suggestion
    command: string, args: string[], options: SpawnSyncOptions = {}): SpawnResult {
```

devversion commented on dev-infra/utils/child-process.ts:
```diff
@@ -83,20 +95,51 @@ export function spawnWithDebugOutput(
+
+
+
+/** Convert the provided exitCode and signal to a single status code. */
```
I'd propose we expand this comment a little and mention that a child process `exits` either with a status, or with a signal. Maybe even link to: https://nodejs.org/api/child_process.html#child_process_event_exit

Also it could be simplified/made more readable with:

```ts
return exitCode ?? signal ?? new Error('Unexpected status/signal combination')`. 
```

devversion commented on dev-infra/utils/git/authenticated-git-client.ts:
```diff
@@ -14,6 +14,10 @@ import {GitClient} from './git-client';
+
+// Reexport of `GitCommandError` as AuthenticatedGitClient also throws this error.
+export {GitCommandError} from './git-client';
+
```
I intentionally kept that in the `git-client` class as I don't see anything wrong with importing directly from there, especially since it's known that `AuthenticatedGitClient` is a superclass of `GitClient`.

devversion commented on dev-infra/utils/git/git-client.ts:
```diff
@@ -24,6 +23,7 @@ export class GitCommandError extends Error {
     // accidentally leak the Github token that might be used in a command,
     // we sanitize the command that will be part of the error message.
     super(`Command failed: git ${client.sanitizeConsoleOutput(args.join(' '))}`);
+    Object.setPrototypeOf(this, GitCommandError.prototype);
```
I assume it's for `instanceof` and that ES5 issue we had in other places too. I wonder though why this is needed just now; we had used `instanceof GitCommandError` in a couple of places.

devversion reviewed (approved):
LGTM. A couple of minor comments

gkalpak reviewed (approved):
🚀

AndrewKushnir commented:
FYI, I was unable to merge this PR, see the error below:
```
Auto-merging dev-infra/utils/git/git-client.ts
Auto-merging dev-infra/pr/discover-new-conflicts/index.ts
CONFLICT (content): Merge conflict in dev-infra/pr/discover-new-conflicts/index.ts
Auto-merging dev-infra/ng-dev.js
CONFLICT (content): Merge conflict in dev-infra/ng-dev.js
error: could not apply 2cf16e03dc... fixup! refactor(dev-infra): add spawnSync to child process utils, normalize naming of child-process utils
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 2cf16e03dc... fixup! refactor(dev-infra): add spawnSync to child process utils, normalize naming of child-process utils
Previous HEAD position was 390d32a876 refactor(dev-infra): add spawnSync to child process utils, normalize naming of child-process utils
Switched to branch 'master'
ng-dev pr merge <pr>

Merge a PR into its targeted branches.

Positionals:
  pr  The PR to be merged.                                                                           [number] [required]

Options:
  ...

Error: Command failed: git rebase --interactive --autosquash 762e057e85fe92149d48aa0c5bc9168147d45d5c merge_pr_head
    at new GitCommandError (/dev-infra/ng-dev.js:317:16)
    at AuthenticatedGitClient.GitClient.run (/dev-infra/ng-dev.js:345:19)
    at AutosquashMergeStrategy.<anonymous> (/dev-infra/ng-dev.js:4171:34)
   ...
```
Adding the "cleanup" label for now...

AndrewKushnir commented:
Quick update: tried to merge to the master branch only (by changing labels in this PR), the result is the same...
